src/hardware/serialport/softmodem.cpp

```diff
@@ -29,6 +29,75 @@
 #include "softmodem.h"
 #include "misc_util.h"
 
+static const char phoneValidChars[] = "01234567890*=,;#+>";
+
+class CPhonebookEntry {
+public:
+	CPhonebookEntry(const char *_phone, const char *_address) {
+		safe_strncpy(phone, _phone, sizeof(phone));
+		safe_strncpy(address, _address, sizeof(phone));
+	}
+
+	const char *IsMatchingPhone(const char *input) {
+		if (strcmp(input, phone) != 0)
+			return NULL;
+
+		return address;
+	}
+
+	static char IsValidPhone(const char *input) {
+		for (const char *ch = input; *ch; ch++) {
+			if (!strchr(phoneValidChars, *ch)) {
+				LOG_MSG("SERIAL: Phone %s contains invalid character %c", input, *ch);
+				return false;
+			}
+		}
+
+		return true;
```

Comment:
If we take in a string, we can use its `find_first_not_of` function to check for invalid characters. Also suggest returning a `bool` given this is a pass/fail check.

```
	static bool IsValidPhone(const std::string& input) {
		if (input.find_first_not_of(phoneValidChars) != std::string::npos) {
			LOG_MSG("SERIAL: Phone %s contains invalid character %c", input, *ch);
			return false;
		}
		return true;
```